src/org/sosy_lab/common/io/TempDir.java

```diff
@@ -0,0 +1,167 @@
+// This file is part of SoSy-Lab Common,
+// a library of useful utilities:
+// https://github.com/sosy-lab/java-common-lib
+//
+// SPDX-FileCopyrightText: 2021 Dirk Beyer <https://www.sosy-lab.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package org.sosy_lab.common.io;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sosy_lab.common.io.TempFile.TMPDIR;
+
+import com.google.common.base.Strings;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import com.google.errorprone.annotations.Immutable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class TempDir {
+
+  private TempDir() {}
+
+  /** Create a builder for temporary directories. */
+  public static TempDirBuilder builder() {
+    return new TempDirBuilder();
+  }
+
+  public static final class TempDirBuilder {
```

Comment:
If the "delete on JVM exit" feature is removed, is it really worth having `TempDirBuilder`?

It would basically only wrap `new DeleteOnCloseDir(Files.createTempDirectory(dir, prefix, fileAttributes))`. This seems like a simple method would be enough?

(For temp files, the situation is different, because we have several more setters on the builder, so offering all feature combinations explods, and some setters have the same or overlapping parameter types, so there can easily be a confusion of parameters if added to a single method. But both problems do not exist here.)

Reply:
Your argumentation makes sense to me, thank you for the detailed explanation.

I moved `DeleteOnCloseDir` to `TempFile` and added the static method `createDeleteOnCloseDir` with non-null prefix and file attributes as parameters to `TempFile`. Creating a temporary directory with `TempFile.createDeleteOnCloseDir(...)` seems natural to me as I would use the directory for storing temporary files.
Do you agree with that?

Following the move I also extracted the common code which improves the message of the `IOException`s.